algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -476,16 +511,56 @@ impl<E: PairingEngine> VerifierKey<E> {
             .as_ref()
             .and_then(|v| v.binary_search_by(|(d, _)| d.cmp(&degree_bound)).ok().map(|i| v[i].1.clone()))
     }
-}
 
-impl<E: PairingEngine> VerifierKey<E> {
     pub fn max_degree(&self) -> usize {
         self.max_degree
     }
 
     pub fn supported_degree(&self) -> usize {
         self.supported_degree
     }
+
+    pub fn union<'a, T: IntoIterator<Item = &'a Self> + Clone>(verifier_keys: T) -> Self {
+        let mut bounds_seen = HashSet::<usize>::new();
+        let mut bounds_and_neg_powers = vec![];
+        let mut bounds_and_prepared_neg_powers = vec![];
+        let mut biggest_vk = None;
+        for vk in verifier_keys {
+            if biggest_vk.is_none() {
+                biggest_vk = Some(vk);
+            } else if biggest_vk.unwrap().supported_degree() < vk.supported_degree() {
+                biggest_vk = Some(vk);
+            }
+            let new_bounds = vk.degree_bounds_and_neg_powers_of_h.clone().unwrap();
+            let new_prep_bounds = vk.degree_bounds_and_prepared_neg_powers_of_h.clone().unwrap();
```

Comment:
I think we can avoid the `clone` via `as_ref()`. This is probably important because the prepared powers will be quite large.

Reply:
```suggestion
            let new_bounds = vk.degree_bounds_and_neg_powers_of_h.as_ref().unwrap();
            let new_prep_bounds = vk.degree_bounds_and_prepared_neg_powers_of_h.as_ref().unwrap();
```